numpy/core/defchararray.py

```diff
@@ -51,21 +51,43 @@ def _use_unicode(*args):
     Helper function for determining the output type of some string
     operations.
 
-    For an operation on two ndarrays, if at least one is unicode, the
-    result should be unicode.
+    For an operation on two ndarrays, if at least one is a dtype with a
+    scalar that subclasses str or bytes, the result should be that
+    dtype. Otherwise, if at least one is unicode, the result should be
+    unicode.
     """
     for x in args:
-        if (isinstance(x, str) or
-                issubclass(numpy.asarray(x).dtype.type, unicode_)):
+        scalar_type = numpy.asarray(x).dtype.type
+        if (isinstance(x, str) or issubclass(scalar_type, unicode_)):
```

Comment:
This smells a bit like it may have been a fast-path for a scalar unicode?  OTOH, I am not sure that is really worthwhile...

Reply:
The logic ended up a little more complicated for `_use_unicode` but I restored the fast path.